# Running a pre-trained YOLO classification model

Since we don't have bounding boxes (and we can't feasibly label all of the bounding boxes in the timelines needed for this project), we're still going to try to make use of pre-trained models and transfer learning to still improve on our model.

The hope is that since we have a relatively smaller-scale dataset (with only aorund 10,000 tornadic exampels to train on), we might see benefits from freezing early layers of a pre-trained model with a lot more data and training only the later layers. 

Ultralytics has a pre-trained YOLO classification model that might come in handy here - we'll try freezing the early layers and training on that front.

In a previous notebook, we only used one year of data (2014) and saw results that weren't so strong in our test set. In the original paper, even using 2 years of data was good enough to get relatively strong results, so we expected that one year of data should still perform reasonably well.

Still, maybe using two years of data might do reasonably well. Let's take data from both 2014 and 2015 and work with that to see how the pre-trained classification model does when re-trained.

In [1]:
from ultralytics import YOLO

# Load a model
model = YOLO("yolo11n-cls.pt")  # load a pretrained model (recommended for training)

dataset_path = "/mnt/d/tornetExamples/all_2014_and_2015"

# Train the model
results = model.train(data="/mnt/d/tornetExamples/all_2014_and_2015", augment=False, epochs=20, save=False, batch=4, freeze=7, imgsz=480)

New https://pypi.org/project/ultralytics/8.3.38 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.36 🚀 Python-3.9.20 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 3070, 8192MiB)
engine/trainer: task=classify, mode=train, model=yolo11n-cls.pt, data=/mnt/d/tornetExamples/all_2014_and_2015, epochs=20, time=None, patience=100, batch=4, imgsz=480, save=False, save_period=-1, cache=False, device=None, workers=8, project=None, name=train25, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=7, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed

E0000 00:00:1732740225.826891    1340 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1732740225.829911    1340 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics

train: Scanning /mnt/d/tornetExamples/all_2014_and_2015/train... 37168 images, 0 corrupt: 100%|██████████| 37168/37168 [
val: Scanning /mnt/d/tornetExamples/all_2014_and_2015/test... 6448 images, 0 corrupt: 100%|██████████| 6448/6448 [00:00<


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 39 weight(decay=0.0), 40 weight(decay=0.0005), 40 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 480 train, 480 val
Using 8 dataloader workers
Logging results to runs/classify/train25
Starting training for 20 epochs...

      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.40it/s]

                   all       0.89          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.46it/s]

                   all      0.893          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:26<00:00,  5.52it/s]

                   all       0.89          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.44it/s]

                   all      0.895          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:26<00:00,  5.50it/s]

                   all      0.896          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.41it/s]

                   all      0.901          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.44it/s]

                   all      0.899          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.41it/s]

                   all      0.901          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.45it/s]

                   all      0.907          1



      Epoch    GPU_mem       loss  Instances       Size


      10/20     0.185G     0.2074          4        480: 100%|██████████| 9292/9292 [22:43<00:00,  6.81it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      11/20     0.187G     0.2046          4        480: 100%|██████████| 9292/9292 [22:36<00:00,  6.85it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.42it/s]

                   all      0.905          1



      Epoch    GPU_mem       loss  Instances       Size


      12/20     0.187G     0.2023          4        480: 100%|██████████| 9292/9292 [22:41<00:00,  6.83it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.36it/s]

                   all      0.905          1



      Epoch    GPU_mem       loss  Instances       Size


      13/20     0.187G     0.1906          4        480:  29%|██▉       | 2681/9292 [06:37<16:19,  6.75it/s]


KeyboardInterrupt: 

Definitely does a bit better, but still only peaks at around a .907 accuracy on the test set before starting to flatline and decrease in validation AUC (hence why we stopped early here).

We're still not seeing very good results out of the transfer-learned model, unfortunately.